lib/sinatra/base.rb

```diff
@@ -1089,7 +1089,16 @@ def invoke
 
     # Dispatch a request with error handling.
     def dispatch!
-      @params.merge!(@request.params).each { |key, val| @params[key] = val && force_encoding(val.dup) }
+      @params.merge!(@request.params).each do |key, val|
```

Comment:
Can this logic be moved into the `force_encoding` method or its own method?

Reply:
force_encoding should only be responsible for force encoding. I was thinking about moving it into its own method, like dup..., but I couldn't imagine that the change leads to bring happiness to us. Can you elaborate about the motivation?

Reply:
I think extracting a method would help explain the need for the slightly unusual looking code. This all started to avoid a `FrozenError` right?  The code doesn't communicate that. I'm not entirely sure if I understand the `force_encoding` method but take a look at this line:

https://github.com/sinatra/sinatra/blob/7b7043cde3515afde713b4da3e54443ca831d3df/lib/sinatra/base.rb#L1756

It looks like the root issue is that a frozen string responds to `force_encoding`.  I think the following would be a legitimate update to that method:

```
if data.respond_to? :force_encoding
  data = data.dup if data.frozen?
  data.force_encoding(encoding).encode!
```

Reply:
I don't agree with the change on `force_encoding` because of two reasons.

1. `force_encoding` is only responsible for force encoding. In other words, it doesn't mean `dup`ing value, it should be managed in caller place, right?  For example, `force_encoding` is also called in other places such as:
- https://github.com/sinatra/sinatra/blob/7b7043cde3515afde713b4da3e54443ca831d3df/lib/sinatra/base.rb#L1026
- https://github.com/sinatra/sinatra/blob/7b7043cde3515afde713b4da3e54443ca831d3df/lib/sinatra/base.rb#L1314

2. In general, our implementation does not always depend on the returned value. Using `dup` can only effect the returned value. Adopting your idea effects not only this issue, but other non-issue code. For example, [a similar change with your idea](https://github.com/sinatra/sinatra/pull/1479/commits/f016c24cd939c21a5284b7a6db813b2c1b280b35) had already been [tested](https://travis-ci.org/sinatra/sinatra/builds/431585215?utm_source=github_status&utm_medium=notification) in #1479, and got some failures.
Of course, your code won't reproduce the failure, but the change has a possibility that leading the similar issue in the future.

For the above reasons, I wouldn't like to `dup` the value inside `force_encoding`.

---
>I think extracting a method would help explain the need for the slightly unusual looking code. 

Yeah I knew. But I couldn't imagine good name for that, and I don't think this is an unusual looking code. However, if you think so, we need to think about improving the readability. 
I think we have two options for that:

1. Extracting method.
2. Adding comments

What do you think?

Reply:
Shouldn't `force_encoding` guard or handle invalid scenarios? I think the fact that it is called in multiple locations only adds to that need. Are we sure that a `FrozenError` isn't a possible issue for them as well?

Anyway, I think a descriptive method name might be `dup_frozen_params`:

```
def dup_frozen_params(params)
  params.each do |key, val|
    if val.frozen?
      params[key] =
          begin
            val = val.dup
          rescue TypeError
            val
          end
    end
  end
end
```

```
@params.merge!(@request.params)
dup_frozen_params(@params)
force_encoding(@params)
```

I'm starting to wonder if this is necessary at all. What if we just changed the following line:

https://github.com/sinatra/sinatra/blob/7b7043cde3515afde713b4da3e54443ca831d3df/lib/sinatra/base.rb#L1756

```
# This seems like a fair guard right?
if !data.frozen && data.respond_to? :force_encoding
```

Is there a need to force encode copies of frozen params? We could just leave it at that ;)

Reply:
Do you think that your proposal can **handle** with any problem? It is a mistake if it is.
You should focus on the name `force_encoding`.

In the first place, if you think that not to raise an exception equals with **handling** any problem, that is completely wrong. It is better to have an easy-to-understand exception than to prevent an exception from occurring with half-hearted thinking. It's obvious.

>Are we sure that a FrozenError isn't a possible issue for them as well?

I can't understand what you are saying. If force_encoding thinks that it should handle any possible scenario it is completely wrong.
There will be situations where we should use `force_encoding` or not. However, at the moment, at least in this PR, it is wrong to introduce dup into force_encoding to crush the exception in any and all scenarios.

Reply:
No, I did not say that.

If it is invalid to call force_encoding on a particular object, then I think we should account for that. Hence my last suggestion...a really simple fix. Don't dup at all and don't attempt to force_encode frozen strings.

https://github.com/sinatra/sinatra/blob/7b7043cde3515afde713b4da3e54443ca831d3df/lib/sinatra/base.rb#L1756

```
# This seems like a fair guard right?
if !data.frozen && data.respond_to? :force_encoding
```

Reply:
Also, if  that change is made it will be easy to write a method `dup_params` which could be called before attempting pass them to `force_encoding`. That was my main suggestion in that last post.

Reply:
You meant not `frozen` but `frozen?`, right? As I said earlier, it's just skipping `force_encoding`. In other words, it can't force encoding when the value is frozen.
Why did you think the proposal is correct?
It's not same behavior with the current implementation and my proposal.

```ruby
mock_app do
  use Class.new {
    def initialize(app)
      @app = app
    end

    def call(env)
      req = Rack::Request.new(env)
      req.update_param('bar', 'baz'.freeze)
      @app.call(env)
    end
  }
  set :default_encoding ,'ISO-8859-1'
  get '/' do
    params[:bar].encoding.to_s
  end
end
get '/'
assert_equal 'ISO-8859-1', body
```

As you can see, the value won't be forced encoding.

>If it is invalid to call force_encoding on a particular object, then I think we should account for that.

Yes, exactly. It is not possible at this time to predict perfectly what kind of situation can be considered in the future, but it will be possible to create a method other than force_encoding or to make corrections at the caller as this time, for example.
At the very least, your proposal is focused on skipping exceptions, which is not at all rational.

Reply:
>Also, if that change is made it will be easy to write a method dup_frozen_params which could be called before attempting pass params to force_encoding. That was my main suggestion in that last post.

Again, ignoring frozen objects can not be a solution. It is not rational.

Reply:
Ok forget that.

How about something like this for extracting a dup method:

```
    def dup_params!
      @params.each do |key, val|
        @params[key] = begin
         val.dup
        rescue TypeError
         val
        end
      end
    end

    # Dispatch a request with error handling.
    def dispatch!
      @params.merge!(@request.params)
      force_encoding(dup_params!)
```